+++
title = "Tratando erros: the tidy way"
date = "2017-04-20 11:07:31"
categories = ["curso-r"]
+++

<div><p class="text-muted text-uppercase mb-small text-right"> Por <a href="http://curso-r.com/author/julio">Julio</a> 20/04/2017 </p><div id="post-content"> <p>Tratar erros no R &#xE9; importante para identificar problemas nos c&#xF3;digos e evitar retrabalho. Quem nunca rodou um algoritmo pesad&#xED;ssimo que deu errado na &#xFA;ltima itera&#xE7;&#xE3;o? Nesse artigo, veremos como trabalhar com erros no R e a vers&#xE3;o <code>tidy</code> dessas solu&#xE7;&#xF5;es.</p>
<div id="usando-try-e-trycatch" class="section level2"> <p>A forma tradicional de tratar erros no R &#xE9; com a fun&#xE7;&#xE3;o <code>tryCatch()</code>. Essa fun&#xE7;&#xE3;o tem como primeiro argumento uma express&#xE3;o a ser avaliada e argumentos diversos para trabalhar com os erros. A vers&#xE3;o mais compacta do <code>tryCatch()</code> &#xE9; escrita assim:</p>
<pre class="r"><code>tryCatch(sqrt(1), error = function(e) e)
## [1] 1
tryCatch(sqrt(&apos;a&apos;), error = function(e) e)
## &lt;simpleError in sqrt(&quot;a&quot;): non-numeric argument to mathematical function&gt;</code></pre>
<p>O <code>try()</code> &#xE9; uma simplifica&#xE7;&#xE3;o de <code>tryCatch()</code> que assume que n&#xE3;o estamos interessados no erro, mas sim no resultado da fun&#xE7;&#xE3;o quando ela d&#xE1; certo. O c&#xF3;digo abaixo n&#xE3;o trava:</p>
<pre class="r"><code>try(sqrt(1))
## [1] 1
try(sqrt(&apos;a&apos;))
## Error in sqrt(&quot;a&quot;) : non-numeric argument to mathematical function</code></pre>
<p>Existe at&#xE9; mesmo uma vers&#xE3;o quieta do <code>try()</code>, usando o par&#xE2;metro <code>silent =</code>. Quando a express&#xE3;o d&#xE1; um erro, o <code>try()</code> retorna a mensagem de erro de forma invis&#xED;vel, ou seja, sem mostrar explicitamente para o usu&#xE1;rio.</p>
<pre class="r"><code>x &lt;- try(sqrt(&apos;a&apos;), silent = TRUE)
x
## [1] &quot;Error in sqrt(\&quot;a\&quot;) : non-numeric argument to mathematical function\n&quot;
## attr(,&quot;class&quot;)
## [1] &quot;try-error&quot;
## attr(,&quot;condition&quot;)
## &lt;simpleError in sqrt(&quot;a&quot;): non-numeric argument to mathematical function&gt;</code></pre>
</div>
<div id="usando-adverbios-do-purrr" class="section level2"> <p>Hoje em dia, o jeito mais arrumado de tratar erros &#xE9; usando as fun&#xE7;&#xF5;es <code>purrr::possibly()</code> e suas amigas, <code>quietly()</code> e <code>safely()</code>. Note que todas essas palavras s&#xE3;o adv&#xE9;rbios: o objetivo delas &#xE9; alterar o comportamento de outros verbos (outras fun&#xE7;&#xF5;es). Essa forma de pensar nos nomes das fun&#xE7;&#xF5;es (fun&#xE7;&#xF5;es s&#xE3;o verbos, modificadores de fun&#xE7;&#xF5;es s&#xE3;o adv&#xE9;rbios) faz parte do <a href="http://curso-r.com/blog/2017/02/15/2017-02-16-manifesto-tidy/">princ&#xED;pio tidy</a>.</p>
<ul>
<li><code>safely()</code> retorna uma lista com elementos <code>result</code> e <code>error</code>. Quando a fun&#xE7;&#xE3;o n&#xE3;o d&#xE1; erro, <code>error</code> fica igual a <code>NULL</code>. Quando a fun&#xE7;&#xE3;o d&#xE1; erro, <code>error</code> guarda a mensagem de erro e <code>result</code> guarda o valor do par&#xE2;metro <code>otherwise =</code>, que por padr&#xE3;o &#xE9; <code>NULL</code>.</li>
<li><code>possibly()</code> &#xE9; uma vers&#xE3;o mais otimista do <code>safely()</code>, que exige a defini&#xE7;&#xE3;o de <code>otherwise =</code> e n&#xE3;o guarda as mensagens de erro.</li>
<li><code>quietly()</code> n&#xE3;o trata erros (ou seja, ela trava quando d&#xE1; erro), mas guarda informa&#xE7;&#xF5;es sobre <em>warnings</em> e <em>messages</em>.</li>
</ul>
<div class="figure"><span id="fig:unnamed-chunk-5"></span>
<img src="http://curso-r.com/blog/2017-04-09-try_files/figure-html/unnamed-chunk-5-1.png" alt="Admita, voc&#xEA; n&#xE3;o imaginava que teria de pensar em gram&#xE1;tica para programar em R." width="768">
<p class="caption">
Figure 1: Admita, voc&#xEA; n&#xE3;o imaginava que teria de pensar em gram&#xE1;tica para programar em R.
</p>
</div>
<p>Vamos ver as tr&#xEA;s fun&#xE7;&#xF5;es colocadas em pr&#xE1;tica. Como exemplo usaremos a fun&#xE7;&#xE3;o <code>log</code>, que i) retorna um n&#xFA;mero quando a entrada &#xE9; um n&#xFA;mero positivo, ii) d&#xE1; um warning quando a entrada &#xE9; um n&#xFA;mero menor ou igual a zero, e iii) d&#xE1; um erro se a entrada n&#xE3;o &#xE9; um n&#xFA;mero.</p>
<pre class="r"><code>log(10)
## [1] 2.302585
log(-1)
## Warning in log(-1): NaNs produced
## [1] NaN
log(&apos;a&apos;)
## Error in log(&quot;a&quot;): non-numeric argument to mathematical function</code></pre>
<p>Vamos fazer as vers&#xF5;es modificadas de <code>log</code>:</p>
<pre class="r"><code>library(purrr)
safe_log &lt;- safely(log) # outra forma fancy de escrever isso: log %&gt;% safely()
possible_log &lt;- possibly(log, otherwise = &apos;putz&apos;)
quiet_log &lt;- quietly(log)</code></pre>
<p>Vamos mapear os seguintes elementos nessas fun&#xE7;&#xF5;es:</p>
<pre class="r"><code>entradas &lt;- list(10, -1, &apos;a&apos;)</code></pre>
<p>Agora, os resultados:</p>
<pre class="r"><code>## Esse c&#xF3;digo vai travar
map(entradas, log) ## Warning in .f(.x[[i]], ...): NaNs produced
## Error in .f(.x[[i]], ...): non-numeric argument to mathematical function</code></pre> <pre class="r"><code>## Retorna uma lista com erros e resultados NULL
map(entradas, safe_log)
## Warning in .f(...): NaNs produced
## [[1]]
## [[1]]$result
## [1] 2.302585
## ## [[1]]$error
## NULL
## ## ## [[2]]
## [[2]]$result
## [1] NaN
## ## [[2]]$error
## NULL
## ## ## [[3]]
## [[3]]$result
## NULL
## ## [[3]]$error
## &lt;simpleError in .f(...): non-numeric argument to mathematical function&gt;</code></pre> <pre class="r"><code>## Retorna uma lista com os resultados que deram certo
map(entradas, possible_log)
## Warning in .f(...): NaNs produced
## [[1]]
## [1] 2.302585
## ## [[2]]
## [1] NaN
## ## [[3]]
## [1] &quot;putz&quot;</code></pre> <pre class="r"><code>## Tamb&#xE9;m trava, mesmo problema de log
map(entradas, quiet_log) ## Error in .f(...): non-numeric argument to mathematical function ## Quando funciona, retorna todos os warnings e messages
map(entradas[1:2], quiet_log)
## [[1]]
## [[1]]$result
## [1] 2.302585
## ## [[1]]$output
## [1] &quot;&quot;
## ## [[1]]$warnings
## character(0)
## ## [[1]]$messages
## character(0)
## ## ## [[2]]
## [[2]]$result
## [1] NaN
## ## [[2]]$output
## [1] &quot;&quot;
## ## [[2]]$warnings
## [1] &quot;NaNs produced&quot;
## ## [[2]]$messages
## character(0)</code></pre>
</div>
<div id="combo-com-purrr-e-tibble" class="section level2"> <p>Um combo que eu gosto bastante de usar &#xE9; adicionar erros do c&#xF3;digo retornado por <code>safely()</code> dentro de uma <code>tibble</code>. Uma forma de fazer isso &#xE9; com o c&#xF3;digo abaixo:</p>
<pre class="r"><code>library(tibble) # adiciona um resultado default caso d&#xEA; erro. No caso, NA.
safe_log2 &lt;- log %&gt;% safely(otherwise = NA_real_) d_result &lt;- entradas %&gt;% map(safe_log2) %&gt;% # ao inv&#xE9;s de uma lista de tamanho 3 com 2 elementos (result e error), # temos uma lista de tamanho 2 (result e error) com 3 elementos. transpose() %&gt;% # simplifica o vetor de resultados simplify_all() %&gt;% # converte para data frame as_tibble() d_result
## # A tibble: 3 x 2
## result error
## &lt;dbl&gt; &lt;list&gt;
## 1 2.302585 &lt;NULL&gt;
## 2 NaN &lt;NULL&gt;
## 3 NA &lt;S3: simpleError&gt;</code></pre>
<p>Assim, &#xE9; poss&#xED;vel guardar as informa&#xE7;&#xF5;es dos erros de forma concisa, sem perder a informa&#xE7;&#xE3;o dos erros.</p>
<p>E &#xE9; isso. Happy coding ;)</p>
<p><strong>PS</strong>: At&#xE9; pouco tempo atr&#xE1;s eu usava a fun&#xE7;&#xE3;o <code>dplyr::failwith()</code> para fazer o mesmo que <code>possibly()</code>. Por&#xE9;m, descobri que essa fun&#xE7;&#xE3;o <a href="https://github.com/tidyverse/dplyr/blob/master/NEWS.md#deprecated-and-defunct">ser&#xE1; retirada do <code>dplyr</code> no futuro</a>. Ent&#xE3;o se voc&#xEA; &#xE9; um usu&#xE1;rio de <code>failwith()</code>, est&#xE1; na hora de mudar!</p>
</div> </div></div>
